### PR TITLE
Add basic compilation facilities, don't relabel qubits

### PR DESCRIPTION
**Pull request by karalekas**
While trying to gather QCS data using mitiq, I remembered (the hard way) that there is no way to "just nativize" when using the Quil compiler, and thus we need a facility to perform nativization without optimization. Fortunately, `forest-benchmarking` had a partially-implemented utility for this (but rather than depend on it, which itself has a heavy set of dependencies, I chose to copy it over). Unfortunately, we lose a lot of the really powerful features that quilc offers by doing this, including automatic qubit rewiring. This is a draft PR because I'm going to test it on QCS to see if it works as intended. If this ends up being a nice tool, might try to get it merged into pyquil.

Additionally updates `to_quil` to keep the qubit labeling the same (which does not happen by default in Cirq) using a little `QubitOrder` hack. Maybe this could end up in Cirq?

Together, these two enhancements should allow us to perform error mitigation on QCS.

### CODE REVIEW

**Comment by karalekas** [comment]
After some more thought, this should be included in the current milestone as we're likely going to use `basic_compile` in executor snippets in the preprint, and those snippets should only contain stuff from released versions of mitiq.

File: mitiq/mitiq_pyquil/tests/test_pyquil_compiler.py [hunk 1]
```diff
@@ -0,0 +1,319 @@
...
+    _Z,
+    match_global_phase,
+)
+
+try:
+    from pyquil.simulation.tools import program_unitary
+
+    unitary_tools = True
+except ImportError:
+    unitary_tools = False
```
Comment (willzeng) [inline mitiq/mitiq_pyquil/tests/test_pyquil_compiler.py:73]: Why is this needed? Is this to support different pyquil versions?

File: mitiq/mitiq_pyquil/compiler.py [hunk 2]
```diff
@@ -0,0 +1,343 @@
...
+    """
+    p = Program()
+    p += RZ(np.pi, q)
+    return p
+
+
+# This function is taken from cirq. License: apache 2.
+def match_global_phase(
+    a: np.ndarray, b: np.ndarray
+) -> Tuple[np.ndarray, np.ndarray]:
```
Comment (willzeng) [inline mitiq/mitiq_pyquil/compiler.py:219]: Given that we depend on cirq, could we just import this?

File: mitiq/mitiq_pyquil/compiler.py [hunk 3]
```diff
@@ -0,0 +1,343 @@
...
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+"""Provides a utility basic_compile to allow for running on QCS without
+using quilc, as quilc will optimize away unitary folding.
+
+NB: Copied in large part from rigetti/forest-benchmarking (Apache-2.0)
+and modified to support a larger gateset (e.g. CPHASE).
```
Comment (rmlarose) [inline mitiq/mitiq_pyquil/compiler.py:20]:
```suggestion
NB: Copied in large part from rigetti/forest-benchmarking (Apache-2.0) [1]
and modified to support a larger gateset (e.g. CPHASE).

[1] https://github.com/rigetti/forest-benchmarking
```

File: mitiq/mitiq_pyquil/compiler.py (hunk 2 above)
Comment (rmlarose) [reply mitiq/mitiq_pyquil/compiler.py:219]: +1.

File: mitiq/mitiq_pyquil/tests/test_pyquil_compiler.py [hunk 4]
```diff
@@ -0,0 +1,319 @@
...
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+"""NB: Copied in large part from rigetti/forest-benchmarking (Apache-2.0)
+and modified to test a larger gateset.
```
Comment (rmlarose) [inline mitiq/mitiq_pyquil/tests/test_pyquil_compiler.py:17]:
```suggestion
"""Tests for the basic pyQuil compiler.

NB: Copied in large part from rigetti/forest-benchmarking (Apache-2.0) [1]
and modified to test a larger gateset.

[1] [1] https://github.com/rigetti/forest-benchmarking
```

File: mitiq/mitiq_pyquil/tests/test_pyquil_compiler.py [hunk 5]
```diff
@@ -0,0 +1,319 @@
...
+
+try:
+    from pyquil.simulation.tools import program_unitary
+
+    unitary_tools = True
+except ImportError:
+    unitary_tools = False
+
+
+def assert_all_close_up_to_global_phase(
```
Comment (rmlarose) [inline mitiq/mitiq_pyquil/tests/test_pyquil_compiler.py:76]: Alternative: [`cirq.equal_up_to_global_phase`](https://cirq.readthedocs.io/en/stable/generated/cirq.equal_up_to_global_phase.html#cirq.equal_up_to_global_phase)

File: mitiq/mitiq_pyquil/tests/test_pyquil_compiler.py (hunk 1 above)
Comment (karalekas) [reply mitiq/mitiq_pyquil/tests/test_pyquil_compiler.py:73]: Yeah, and it was copied over from forest-benchmarking -- I'll go ahead and delete it, because we require a later version of pyquil anyways.

File: mitiq/mitiq_pyquil/compiler.py (hunk 2 above)
Comment (karalekas) [reply mitiq/mitiq_pyquil/compiler.py:219]: Good point!

File: mitiq/mitiq_pyquil/tests/test_pyquil_compiler.py (hunk 5 above)
Comment (karalekas) [reply mitiq/mitiq_pyquil/tests/test_pyquil_compiler.py:76]: Nice!

**Comment by karalekas** [comment]
@willzeng yes i could see this ending up in pyquil at some point, but for now i think we should keep it here

**Comment by karalekas** [comment]
Rebased off of master and fixed the `mypy` issues -- should be ready for final review!
